Support line continuance with pipe at the start of a line

KirkMunro opened the pull request:
# PR Summary

Expands line continuance for pipelines to allow lines to continue automatically without backticks with the pipe symbol at the start of a line. This adds to the existing functionality where pipes can be used to continue lines by placing them at the end of a line.

Supported methods to wrap pipelines that existed before this PR:

```powershell
# Wrapping with a pipe at the end of a line
Get-Process | Where CPU | Where Path |
    Get-Item | Where FullName -match "AppData" | 
    Sort FullName -Unique
```

```powershell
# Wrapping with a backtick at the end of a line and a pipe at the beginning of a line
Get-Process | Where CPU | Where Path `
    | Get-Item | Where FullName -match "AppData" `
    | Sort FullName -Unique
```

Additional supported methods to wrap pipelines with this PR in place:

```powershell
# Wrapping with a pipe at the beginning of a line (no backtick required)
Get-Process | Where CPU | Where Path
    | Get-Item | Where FullName -match "AppData"
    | Sort FullName -Unique
```

```powershell
# Wrapping with a pipe on a line by itself (this comes automatically with
# the addition of parser support for wrapping with a pipe at the beginning
# of a line since PowerShell already supported wrapping with a pipe at the
# end of a line)
Get-Process | Where CPU | Where Path
    |
    Get-Item | Where FullName -match "AppData"
    |
    Sort FullName -Unique
```

## PR Context

See #3020.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [X] Issue filed: [3746](https://github.com/MicrosoftDocs/PowerShell-Docs/issues/3746)
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

KirkMunro committed:
support continuance with pipe at beginning of line

KirkMunro committed:
token management logic changed to fix highlighting

KirkMunro committed:
Merge branch 'master' into line-continuance-with-pipe

KirkMunro committed:
cleaned up new comments

KirkMunro committed:
added pester tests

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,247 @@
+        function ExecuteCommand {
+            param ([string]$command)
+            [powershell]::Create().AddScript($command).Invoke()
+        }
```
Do we need cleanup?
```powershell
        function ExecuteCommand {
            param ([string]$command)
            try {
                $PowerShell.AddScript($command).Invoke()
            }
            finally {
                $PowerShell.Commands.Clear()
            }
        }
```

KirkMunro replied:
Maybe. I wasn't sure. I saw that function, but wasn't sure which is the current/best way to write these pester tests, and didn't want to take a dependency that I didn't need. FWIW, I disagree with the logic used to write the other function -- a new PowerShell instance should be created for every invocation rather than clearing the command stack between invocations of commands (I have found in practice while working with PowerShell from C# a lot that this can lead to trouble and should be avoided, especially when validating tests). If we want this to be cleaned up, I would prefer changing the other tests to use my function than changing my tests to use the other function.

iSazonov commented on test/powershell/Language/Parser/Parsing.Tests.ps1:
```diff
@@ -270,6 +270,7 @@ Describe 'splatting parsing' -Tags "CI" {
 }
 
 Describe 'Pipes parsing' -Tags "CI" {
+    ShouldBeParseError '|gps' EmptyPipeElement 0
```
Do we need
```powershell
ShouldBeParseError '|' EmptyPipeElement 0
```

KirkMunro replied:
The other two tests validate that something is before/after the pipe, and in a better way (it is clear what is missing/non-functional if one of those tests fail). With the proposed test, it would not be clear what became broken should it fail, so I feel it would add more harm than good.

iSazonov commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -362,7 +362,7 @@ private void V3SkipNewlines()
             {
                 _ungotToken = null;
-                _tokenizer.SkipNewlines(false, true);
+                _tokenizer.SkipNewlines(NewlineSkipOption.V3);
```
Please look NoteV3FeatureUsed() - I think we can remove V3.

lzybkr replied:
Yes, this method can be removed. The original intent was to be helpful for a compatibility checker. Features like new keywords would be obvious, but places where PowerShell was more permissive with newlines would have been much harder to detect.

That compatibility checker was never built and nobody cares about V2 to V3 compat anymore.

---
In reply to: [258930501](https://github.com/PowerShell/PowerShell/pull/8938#discussion_r258930501) [](ancestors = 258930501)

KirkMunro replied:
Agreed. I didn't want to touch more than I needed to, but that logic lead to commented out code that is no longer used, so I just cleaned all of that up. Thank you for suggesting I do so.

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,247 @@
+    Context 'Lines ending with a backtick that do not parse' {
+        It 'Lines ending with a single backtick followed immediately by a comment' {
+            $script = @'
+'Hello' + `# This is not a valid comment
```
I'd prefer it works.

vexx32 replied:
This isn't a change from this PR though, is it?Backtick before a comment-start token escapes it. :)

KirkMunro replied:
Correct. I was looking for line continuance tests, and there weren't any, so I put these in this PR so that we have something to start with.

Also, backtick is an escape character, so in this case I think the behavior is absolutely correct, since the backtick escapes the #. Forcing users to add a space here isn't a big deal. Better that we make it so that backticks hardly ever need to be used, other than intentionally escaping a character, and I think we're almost there (missing automatic line continuance for parameters on newlines, so that will come in another PR, but with an experimental flag since it would technically be a breaking change for commands that start with "-").

Jaykul replied:
An example that _would_ parse would be: ```Write-Host Hello `# World```

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,247 @@
+        It 'Lines ending with a pipe followed by whitespace' {
+            $script = @'
+# The next line ends with trailing whitespace
+'Hello' |     
```
End spaces can be auto removed in editors. We should protect. Please use operators to construct the test string.

KirkMunro replied:
Good point, done.

iSazonov replied:
I don't see your update.

KirkMunro replied:
See https://github.com/PowerShell/PowerShell/pull/8938/commits/ad12433f43edee02090efb1b4ef17ac97c61e9fd#diff-6d2b8b1bd72c210d3be378d71e27a146.

iSazonov replied:
It is fix for line 25 - line 82 is not fixed.

KirkMunro replied:
Sorry, missed that one. Commit coming soon.

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,247 @@
+
+        It 'Lines ending with two pipes' {
+            $script = @'
+'Hello' ||
```
Should we check:
```powershell
           $script = @' 
'Hello' |
        | 
'@

KirkMunro replied:
Yes, that is a better test. The other test is necessary, but it's not a line continuance test, so I moved it to Parsing.Tests.ps1.

iSazonov reviewed (commented):
Thanks for great contribution!

KirkMunro committed:
CodeFactor changes

KirkMunro committed:
Removed legacy V3 code that did nothing

KirkMunro committed:
updated pester tests

lzybkr commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -934,6 +956,32 @@ internal void SkipNewlines(bool skipSemis, bool v3)
+                        while (_currentIndex > startIndex)
+                        {
+                            UngetChar();
+                        }
```
Have you compared performance? I have a feeling this will be a hotspot in comment-heavy code.

lzybkr replied:
IIRC, you might also see extra memory usage (compared to before this change) from the rescan when re-parsing a script for it's help because the comment tokens will be created twice.

---
In reply to: [259051896](https://github.com/PowerShell/PowerShell/pull/8938#discussion_r259051896) [](ancestors = 259051896)

KirkMunro replied:
Thanks @lzybkr. I haven't compared performance yet, but noticed in the comments that you (or someone on the PS team) seems to have done some significant performance testing using files from Poshcode, Technet, etc. with comments. Are those tests available somewhere?

Wrt both performance and memory usage, I was actually working on refactoring part of the logic partly because CodeFactor was complaining about the complexity of the SkipNewlines method (which I wasn't quite happy with either), and partly because I didn't like the tokens getting created unnecessarily just for lookahead logic (that's the only memory impact, I believe). I also didn't like using a while loop to move the character marker, and instead simply set it back to where it was if the lookahead fails. I just need to figure out why this new logic isn't syntax highlighting the second `gcm`command in this script, and add a test to make sure that works properly:

```
gcm Get-Command
gcm Get-Help
```

iSazonov replied:
> Are those tests available somewhere?

Jason put some helpers in tools/performance folder.

lzybkr replied:
The corpus you've seen reference to includes many internal scripts that can't be shared and honestly those scripts were maybe the most valuable in that corpus.

Today I'd look to collecting scripts from the PowerShell Gallery and GitHub. (Google BigQuery reports >400,000 scripts on GitHub, though it looks like it reports many duplicates from forks despite the claim that they exclude forks.)

iSazonov replied:
Another question. For backtick we create `NewToken(TokenKind.LineContinuation)` - should we create the same for automatic line continuation?

lzybkr replied:
I'm not sure. Some tools (e.g. a code formatter) might insert the backtick.

iSazonov replied:
For consistency it could be `TokenKind.AutoLineContinuation`.

Jaykul replied:
Yeah, anyone want to put in the effort to collect a safe subset of scripts from the internet for performance testing? 😉 

I mean, most scripts on github are likely to be build scripts (not very good for perf testing of PowerShell, probably) or actually just functions that are part of a module -- but there are still tens of thousands of [PowerShell gists](https://gist.github.com/search?utf8=%E2%9C%93&q=language%3APowerShell+if&ref=searchresults) too ...

powercode replied:
@Jaykul I think we only need to test the parser performance in this case - not actually running them.

Jaykul replied:
Very good point 😊

KirkMunro replied:
I'll run a bunch of tests once I get back to this PR. I decided to put the finishing touches on tests for my PR with the `S.M.A.PowerShell.InvokeAsync` methods first to clear my head a little before digging in deeper in tests with this PR.

iSazonov replied:
I added WIP until we get new commits.

KirkMunro replied:
Ok, I've moved forward on this PR, and have a few comments:

1. @iSazonov: I don't think we should have a token for automatic line continuation. A backtick is a physical token. All other tokens are physical tokens in a script. In this case (automatic line continuation), there is no token to capture, so I think having a new token where there is nothing in the script would just confuse things.
2. @lzybkr: I completely refactored the logic I was using and now just lookahead without tokenizing to determine if there is continuation or not. That fixed the tokenization problem I couldn't put my finger on, plus it keeps memory constant, and it's very fast. Given the change in code, I would like you to have another look. Initial performance tests show no noticeable impact in performance (tested parsing the allcommands.ps1 file from the dbatools module, using the v2 parser and the S.M.A.L.Parser ParseFile method).

KirkMunro replied:
Ok, I've moved forward on this PR, and have a few comments:

1. @iSazonov: I don't think we should have a token for automatic line continuation. A backtick is a physical token. All other tokens are physical tokens in a script. In this case (automatic line continuation), there is no token to capture, so I think having a new token where there is nothing in the script would just confuse things.
2. @lzybkr: I completely refactored the logic I was using and now just lookahead without tokenizing to determine if there is continuation or not. That fixed the tokenization problem I couldn't put my finger on, plus it keeps memory constant, and it's very fast. Given the change in code, I would like you to have another look. Initial performance tests show no noticeable impact in performance (tested parsing the allcommands.ps1 file from the dbatools module, using the v2 parser and the S.M.A.L.Parser ParseFile method). I'm working out more performance tests now.

KirkMunro replied:
More performance testing results:

I created a 77MB script file with a lot of block comments containing Lorem Ipsum text and line comments, with no blank lines (intended to be an extreme/worst case scenario based on the lookahead logic), and ran it through the parser (new and 2.0) before and after these changes. Before the changes I get parse times around 8.7s for the file, regardless of which parser I use. After the changes I get parse times ranging from 9.4s up to 10s.

I then created a simple function with the example comment-based help that is in the docs at the end of the function, and copied/pasted the function hundreds of times in the same file to create another 77MB script file, and ran it through the parser (new and 2.0) before and after these changes, and timed the parsers using Measure-Object. The times were identical in this case.

One thought I'd like an opinion on that would reduce the likelihood of this causing any performance issue whatsoever for anyone: with backtick and pipe at the end of the line, we have an explicit line continuance, and can continue our command even after multiple blank lines. With "implicit" (for lack of a better term) line continuance like this, currently it works the same way, allowing scripters to apply multiple blank lines if they choose to do so. Should it? Or should we have the parser enforce that no blank lines or lines consisting of nothing but whitespace are allowed with implicit/automatic line continuance? I think it is very unlikely that someone would want a blank line when using this -- comments, sure, but not blank lines. In testing (I already threw together an experimental build that identifies blank lines or lines containing nothing but whitespace as command breaks when it comes to automatic line continuance), I took the same scrpit I used in the first file and generated a new file with blank lines before the block comments. The result was that pre-automatic line continuance, the file parses in around 8.8s, and post-automatic line continuance, the file parses in around 9.4s, which is a nice improvement over the results of the first test performed above.

@lzybkr, @iSazonov: Any thoughts on whether or not blank/empty lines should allow automatic line continuance or not?

iSazonov replied:
> With "implicit" (for lack of a better term) line continuance like this, currently it works the same way, allowing scripters to apply multiple blank lines if they choose to do so. Should it? 

It is tricky question :-) We have different behavior with PSReadline and without it. Perhaps @mklement0  would be interested to see it. Personally, I think PSReadline behavior is more correct.

As for performance tests, the use of Measure-Command does not give a reliable result, although it clearly shows the absence of performance degradation. We could get a more accurate result using WPR/WPA, but it only makes sense after moving to .Net Core 3.0. I hope to find time to drive more carefully the new code, but for now I like the result obtained and I think we could merge before moving to .Net Core 3.0.

KirkMunro replied:
I agree that the PSReadline behaviour is the correct one. I'm not sure my original question was clear though. When applying automatic line continuance, should we allow blank/whitespace lines in a command or not.

e.g. If someone enters this command, should it work?

```PowerShell
# Get the current process
gps -id $pid

# And pull out its name
| foreach Name
```

Or, should that only work when there is no blank/whitespace line in the middle (comment lines are allowed though), like this:

```PowerShell
# Get the current process
gps -id $pid
# And pull out its name
| foreach Name
```

Mulling this over on the weekend I've found myself thinking blank lines should not be allowed, to keep multi-line commands contiguous in scripts where they are used. I'm also hoping to put in an experimental feature allowing parameter wrapping, and for that I wouldn't want blank lines used either.

e.g. Parameter wrapping (experimental since it would be breaking for the unusual case where you have a command that starts with "-", requiring you to invoke such commands with the call operator).

```PowerShell
Get-Command
    -Name Invoke-Command
    -Syntax
    @AdditionalSplattedParametersIfDesired
```

That's also a contrived example, but the point is to offer easier line wrapping for commands without requiring users to use splatting unless they want/need to.

fourpastmidnight replied:
I'd say that blank lines should not be allowed in implicit command continuations. It smells to me of the classic `if`/`else` statements without braces problem in other languages:

```csharp
if (!something)
    // DoThisThingToDealWithIt(); 
DoTheNextThing();
// Oops...
```

Anyway, I'm super excited to see this feature make it into PowerShell. It's annoyed me for a long time that a command cannot be continued with a pipe on a new line, letting the code line up nicer and look as nice as F#.

iSazonov replied:
My thoughts is the same - extra lines is a way to tricky errors.

KirkMunro committed:
Merge branch 'master' into line-continuance-with-pipe

KirkMunro committed:
made whitespace at end of line intentional

KirkMunro committed:
refactored for performance and to fix syntax highlighting

KirkMunro committed:
Merge branch 'master' into line-continuance-with-pipe

KirkMunro committed:
CodeFactor changes

KirkMunro committed:
removing unnecessary/unintentional change

iSazonov commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -1294,6 +1305,65 @@ private bool OnlyWhitespaceOrCommentsAfterExtent(InternalScriptExtent extent)
+
+                if (_script[i] == '\n' || _script[i] == '\r')
+                {
+                    i = SkipNewline(i);
```
In SkipNewline we again check '\n' and `\r`.
If we accept "_script.Length - 1" from previous comment we can simplify the skip.

KirkMunro replied:
I inlined the skip to avoid additional redundant checks.

iSazonov commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -1294,6 +1305,65 @@ private bool OnlyWhitespaceOrCommentsAfterExtent(InternalScriptExtent extent)
+                {
+                    if (_script[i + 1] == '\n' || _script[i + 1] == '\r')
+                    {
+                        i = SkipNewline(i + 1);
```
The same (double check).

KirkMunro replied:
I inlined the skip to avoid additional redundant checks.

KirkMunro commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -371,7 +357,7 @@ private void SkipNewlinesAndSemicolons()
             {
                 _ungotToken = null;
-                _tokenizer.SkipNewlines(true, false);
+                _tokenizer.SkipNewlines(true);
```
Since SkipNewlines now has one and only one flag, is it really necessary to use the argument name here?

iSazonov replied:
It is our common pattern (to make code more readable).

KirkMunro commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -1294,6 +1305,65 @@ private bool OnlyWhitespaceOrCommentsAfterExtent(InternalScriptExtent extent)
+        {
+            // If the first non-newline, non-whitespace, non-comment, non-backtick, non-semi-colon character
+            // following the newline is a pipe, we have a pipe continuance.
+            for (int i = extent.EndOffset; i < _script.Length;)
```
To make sure I understand, in addition to the change above you are proposing I change line 1364 to `return _script[i + 1] == '|';`? For what benefit? I reviewed this proposal, and feel the change would make the logic more difficult to understand than its current implementation.

iSazonov replied:
Benefit is that we remove two boundary checks (`(i + 1) < _script.Length`) in the cycle.
Below example. Also I moved checks for space and new line at first position (frequent cases) and use temporary variable (I expect better code gen specially if we unroll SkipNewline).
```c#
        private bool PipeContinuanceAfterExtent(InternalScriptExtent extent)
        {
            // If the first non-newline, non-whitespace, non-comment, non-backtick, non-semi-colon character
            // following the newline is a pipe, we have a pipe continuance.
            for (int i = extent.EndOffset; i < (_script.Length - 1);)
            {
                char c = _script[i];

                if (char.IsWhitespace(c))
                {
                    i++;
                    continue;
                }

                if (c == '\n' || c == '\r')
                {
                    i = SkipNewline(i);
                    continue;
                }

                if (c == '#')
                {
                    // SkipLineComment will return the position after the comment end
                    // which is either at the end of the file, or a cr or lf.
                    i = SkipLineComment(i + 1);
                    continue;
                }

                if (c == '<' && _script[i + 1] == '#')
                {
                    i = SkipBlockComment(i + 2);
                    continue;
                }

                if (c == '`')
                {
                    if (_script[i + 1] == '\n' || _script[i + 1] == '\r')
                    {
                        i = SkipNewline(i + 1);
                        continue;
                    }

                    if (char.IsWhiteSpace(_script[i + 1]))
                    {
                        i += 2;
                        continue;
                    }
                }

                return c == '|';
            }

            return _script[_script.Length - 1] == '|';
        }
```

KirkMunro replied:
That was helpful. Thanks @iSazonov. I have made these changes, plus inlined the SkipNewline function.

iSazonov committed:
Add argument qualifier to make code more readable

Co-Authored-By: KirkMunro <KirkMunro@users.noreply.github.com>

iSazonov committed:
Add argument qualifier to make code more readable

Co-Authored-By: KirkMunro <KirkMunro@users.noreply.github.com>

KirkMunro committed:
refactor based on PR feedback

KirkMunro committed:
automatic continuance requires contiguous code

KirkMunro committed:
Merge branch 'master' into line-continuance-with-pipe

KirkMunro commented:
Removed WIP tag. Pushed logic that requires code to be contiguous for automatic line continuance to work. Updated Pester tests to check for multiple types of whitespace (identified from code), and added new test to ensure that automatic line continuance with pipe on the start of a line only works when lines are contiguous.

This is PR now ready for final review/merge.

KirkMunro committed:
Merge branch 'master' into line-continuance-with-pipe

KirkMunro committed:
CodeFactor changes

iSazonov commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -874,11 +874,10 @@ internal static bool IsKeyword(string str)
-        // so we create them when asked to create them.
-
+            // We normally don't create any tokens here, but the V2 tokenizer api returns newline tokens,
+            // so we create them when asked to create them.
```
It seems the comment is out of date - currently we can create Newline, Semi and LineContinuation tokens.

KirkMunro replied:
Agreed, I'll remove it.

KirkMunro replied:
Updated.

iSazonov commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -980,6 +965,32 @@ private void SkipWhiteSpace()
             }
         }
 
+        private void ScanNewline(char c)
```
The name confuse me. We do something only if TokenList != null.

KirkMunro replied:
When I started working on this code, there were a bunch of Scan* and Skip* methods, but the way those verbs were used was not consistent. I normalized this such that Scan* methods are used to scan characters and add new tokens, and Skip* methods are used to skip characters.

The check for `TokenList != null` was there before this PR, so I kept it, assuming it was necessary. Since you called it out, I'm now curious about what might cause the situation where a Scan* method is invoked when TokenList is null. I'll think about this and check it when I get back to my Windows laptop where I have that PR already to test/debug.

iSazonov replied:
Maybe @daxian-dbw have ideas how better name the methods.

KirkMunro replied:
Yeah, since we're discussing this, it seems like a bug to me as well, because the method would be called and if TokenList was null, do nothing. Not advance the character marker, not add the token. That logic was in place before, so this is a legacy problem. I'm at a conference right now (PowerShell and DevOps Summit), so distracted, but I will revisit that code soon.

KirkMunro replied:
Regarding TokenList, I kept the condition so that memory optimizations remain (it was there before), but I modified the logic so that only the token creation is skipped if TokenList is null (this eliminates what felt like a bug where CR in CRLF wouldn't be skipped if TokenList was null).

KirkMunro replied:
Regarding the name of the method, I went with NormalizeCRLF, since that's exactly what the method does. SkipCR was not appropriate because it only skips CRs that are followed by LF.

iSazonov commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -980,6 +965,32 @@ private void SkipWhiteSpace()
+            }
+        }
+
+        private void ScanSemicolon()
```
The same. The name confuse me. We do something only if TokenList != null.

KirkMunro replied:
The TokenList check has been moved slightly, with a comment explaining why it is there. The name stays the same.

iSazonov reviewed (approved):
LGTM with 3 minor comment.

lzybkr reviewed (commented):
Changes look pretty good, just some minor feedback.

KirkMunro commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -1079,7 +1090,7 @@ internal void ReplaceSavedTokens(Token firstOldToken, Token lastOldToken, Token
         }
 
-        private void ScanNewline(char c)
+        private void SkipNewline(char c)
```
I saw that as well, but hadn't come to a decision on how to better represent/name it. After reading your comment, now I'm liking NormalizeLineEnding as a more appropriate method name, but I'll think about it a little more.

iSazonov replied:
For notice, below Jason use `MaybeSkipCR()` name. I think we could use short `SkipCR()` name with method description "normalize line endings to a single character".

KirkMunro replied:
See previous comment, the method was renamed as `NormalizeCRLF()`.

vexx32 commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -1684,7 +1785,7 @@ private void ScanBlockComment()
                 if (c == '\r' || c == '\n')
                 {
-                    ScanNewline(c);
+                    SkipNewline(c);
```
Wouldn't it be better to ditch the empty if statement and have the second just be `if (c != '\n') {}`?

lzybkr replied:
Probably - I was also looking at other uses where the body wouldn't be empty, so was suggesting the general pattern.

KirkMunro replied:
Refactored that so that the logic is cleaner after taking these suggestions into account.

KirkMunro committed:
Merge branch 'master' into line-continuance-with-pipe

KirkMunro committed:
changes for PR feedback; added some Pester tests

KirkMunro commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -1320,6 +1331,96 @@ private bool OnlyWhitespaceOrCommentsAfterExtent(InternalScriptExtent extent)
+                    i++;
+                    continue;
+                }
+                else if (c == '\r' && _script[i + 1] == '\n')
```
Fixed. Thanks for catching that. I also added specific Pester tests that just use `\r` to ensure it remains fixed.

KirkMunro commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -1320,6 +1331,96 @@ private bool OnlyWhitespaceOrCommentsAfterExtent(InternalScriptExtent extent)
+            // following the newline is a pipe, we have a pipe continuance.
+            bool lastNonWhitespaceIsNewline = true;
+
+            for (int i = extent.EndOffset; i < _script.Length - 1;)
```
Agreed. It was a `for` loop because that made sense, until the logic was changed during the review process, and `while` is a better choice now, so I made that change. I also added a comment explaining the `-1`.

KirkMunro commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -1320,6 +1331,96 @@ private bool OnlyWhitespaceOrCommentsAfterExtent(InternalScriptExtent extent)
+                        i += 2;
+                        continue;
+                    }
+                    else if (c2 == '\r' && i + 2 < _script.Length && _script[i + 2] == '\n')
```
Thanks for catching, fixed, and added Pester tests to ensure it stays fixed.

KirkMunro commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -1320,6 +1331,96 @@ private bool OnlyWhitespaceOrCommentsAfterExtent(InternalScriptExtent extent)
 
+        internal bool IsPipeContinuance(IScriptExtent extent)
+        {
+            var scriptExtent = (InternalScriptExtent)extent;
```
Good catch, that wasn't necessary. Fixed.

iSazonov commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -981,6 +966,38 @@ private void SkipWhiteSpace()
+        {
+            _tokenStart = _currentIndex - 1;
+            NormalizeCRLF(c);
+            // Memory optimization: only create the token if it will be stored
```
```suggestion

            // Memory optimization: only create the token if it will be stored
```

iSazonov commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -981,6 +966,38 @@ private void SkipWhiteSpace()
+        private void ScanSemicolon()
+        {
+            _tokenStart = _currentIndex - 1;
+            // Memory optimization: only create the token if it will be stored
```
```suggestion

            // Memory optimization: only create the token if it will be stored
```

iSazonov commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -981,6 +966,38 @@ private void SkipWhiteSpace()
+        {
+            _tokenStart = _currentIndex - 2;
+            NormalizeCRLF(c);
+            // Memory optimization: only create the token if it will be stored
```
```suggestion

            // Memory optimization: only create the token if it will be stored
```

iSazonov commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -1080,8 +1097,9 @@ internal void ReplaceSavedTokens(Token firstOldToken, Token lastOldToken, Token
         }
 
-        private void ScanNewline(char c)
+        private void NormalizeCRLF(char c)
```
What about `NormalizeEOL()`?

KirkMunro replied:
Not specific enough. There are multiple possible end of lines. This method specifically normalizes CRLF line endings.

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,280 @@
+'Hello' + `
+    ' world'
+'@
+            ExecuteCommand $script | Should -Be 'Hello world'
```
```suggestion
            ExecuteCommand $script | Should -BeExactly 'Hello world'
```

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
````diff
@@ -0,0 +1,280 @@
+
+        It 'Lines ending with a single backtick followed only by a CR (old-style Mac line ending)' {
+            $script = "'Hello' + ```r' world'"
+            ExecuteCommand $script | Should -Be 'Hello world'
````
```suggestion
            ExecuteCommand $script | Should -BeExactly 'Hello world'
```

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,280 @@
+
+    ' world'
+'@
+            ExecuteCommand $script | Should -Be 'Hello world'
```
```suggestion
            ExecuteCommand $script | Should -BeExactly 'Hello world'
```

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,280 @@
+'Hello' + ` # You can place comments after whitespace following backticks
+    ' world'
+'@
+            ExecuteCommand $script | Should -Be 'Hello world'
```
```suggestion
            ExecuteCommand $script | Should -BeExactly 'Hello world'
```

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,280 @@
+# You can place comments in the middle of a continued line
+    ' world'
+'@
+            ExecuteCommand $script | Should -Be 'Hello world'
```
```suggestion
            ExecuteCommand $script | Should -BeExactly 'Hello world'
```

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,280 @@
+'Hello' |
+    ForEach-Object {"$_ world"}
+'@
+            ExecuteCommand $script | Should -Be 'Hello world'
```
```suggestion
            ExecuteCommand $script | Should -BeExactly 'Hello world'
```

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,280 @@
+
+        It 'Lines ending with a pipe followed only by a CR (old-style Mac line ending)' {
+            $script = "'Hello' |`r    ForEach-Object {`"`$_ world`"}"
+            ExecuteCommand $script | Should -Be 'Hello world'
```
```suggestion
            ExecuteCommand $script | Should -BeExactly 'Hello world'
```

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,280 @@
+
+    ForEach-Object {"$_ world"}
+'@
+            ExecuteCommand $script | Should -Be 'Hello world'
```
```suggestion
            ExecuteCommand $script | Should -BeExactly 'Hello world'
```

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,280 @@
+'Hello' |# You can place comments after whitespace following pipes
+    ForEach-Object {"$_ world"}
+'@
+            ExecuteCommand $script | Should -Be 'Hello world'
```
```suggestion
            ExecuteCommand $script | Should -BeExactly 'Hello world'
```

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,280 @@
+# You can place comments in the middle of a continued pipeline
+    ForEach-Object {"$_ world"}
+'@
+            ExecuteCommand $script | Should -Be 'Hello world'
```
```suggestion
            ExecuteCommand $script | Should -BeExactly 'Hello world'
```

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,280 @@
+'Hello'
+    | ForEach-Object {"$_ world"}
+'@
+            ExecuteCommand $script | Should -Be 'Hello world'
```
```suggestion
            ExecuteCommand $script | Should -BeExactly 'Hello world'
```

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,280 @@
+
+        It 'Line continuance using a pipe at the start of a subsequent line after a CR (old-style Mac line ending)' {
+            $script = "'Hello'`r    | ForEach-Object {`"`$_ world`"}"
+            ExecuteCommand $script | Should -Be 'Hello world'
```
```suggestion
            ExecuteCommand $script | Should -BeExactly 'Hello world'
```

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,280 @@
+    # You can place comments before continued pipelines
+    | ForEach-Object {"$_ world"}
+'@
+            ExecuteCommand $script | Should -Be 'Hello world'
```
```suggestion
            ExecuteCommand $script | Should -BeExactly 'Hello world'
```

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,280 @@
+    |
+    ForEach-Object {"$_ world"}
+'@
+            ExecuteCommand $script | Should -Be 'Hello world'
```
```suggestion
            ExecuteCommand $script | Should -BeExactly 'Hello world'
```

iSazonov commented on test/powershell/Language/Parser/LineContinuance.Tests.ps1:
```diff
@@ -0,0 +1,280 @@
+    # You can place comments before continued pipelines
+    ForEach-Object {"$_ world"}
+'@
+            ExecuteCommand $script | Should -Be 'Hello world'
```
```suggestion
            ExecuteCommand $script | Should -BeExactly 'Hello world'
```

KirkMunro committed:
Code review change

Co-Authored-By: Ilya <darpa@yandex.ru>

KirkMunro committed:
Code review change

Co-Authored-By: Ilya <darpa@yandex.ru>

KirkMunro committed:
Code review change

Co-Authored-By: Ilya <darpa@yandex.ru>

KirkMunro committed:
Code review change

Co-Authored-By: Ilya <darpa@yandex.ru>

KirkMunro committed:
Code review change

Co-Authored-By: Ilya <darpa@yandex.ru>

KirkMunro committed:
Code review change

Co-Authored-By: Ilya <darpa@yandex.ru>

KirkMunro committed:
Code review change

Co-Authored-By: Ilya <darpa@yandex.ru>

KirkMunro committed:
Code review change

Co-Authored-By: Ilya <darpa@yandex.ru>

KirkMunro committed:
Code review change

Co-Authored-By: Ilya <darpa@yandex.ru>

vexx32 commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -1321,6 +1342,100 @@ private bool OnlyWhitespaceOrCommentsAfterExtent(InternalScriptExtent extent)
+                    continue;
+                }
+
+                if (c == '`')
```
Given that already as of this PR, the following two scenarios are very different, I don't think there's anything wrong with this.

```powershell
PS> Get-Verb
| % Verb

PS> Get-Verb
-Verb get
```

daxian-dbw replied:
My concern is on the explicit line continuation token.
- First, I don't see why we want to allow backtick when deciding for the pipe continuance. For one-line comment and command block, absolutely, but I don't see a use case where allowing the backtick is useful. It would be great if you have an example.
- Second, the explicit continuation token make things a bit confusing. If `| % verb` on the next line of a backtick works, then a user would assume `-Verb get` on the next line of a backtick works too, because they are following the continuation token.

KirkMunro replied:
I agree with @daxian-dbw. Having backtick support here just makes for poorly written code in this case, and since I changed this to only support continuance when the lines are contiguous with comments in the middle, the backtick support here does more harm than good. Plus, the method is much easier to digest with that removed.

I updated the code with the backtick support removed from that method, and added a specific Pester test to make sure that it is not supported when using implicit line continuance.

KirkMunro replied:
On a related note, the sooner we get this done, the sooner I can add a PR with an experimental feature to support automatic line continuance with parameter names. :) Just waiting for this first since it has significant parser/tokenizer changes.

KirkMunro committed:
Merge branch 'master' into line-continuance-with-pipe

KirkMunro committed:
changes for code review

KirkMunro commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -1321,6 +1342,100 @@ private bool OnlyWhitespaceOrCommentsAfterExtent(InternalScriptExtent extent)
+        private bool PipeContinuanceAfterExtent(IScriptExtent extent)
+        {
+            // If the first non-newline, non-whitespace, non-comment, non-backtick, non-semi-colon character
+            // following the newline is a pipe, we have a pipe continuance.
```
Agreed, and corrected. Thank you.

daxian-dbw reviewed (approved):
LGTM. Thanks @KirkMunro for your contribution!
